### PR TITLE
JAMES-3756 Implement DelegationStore based Authorizator

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/Authorizator.java [hunk 1]
```diff
@@ -30,7 +30,8 @@ public interface Authorizator {
     enum AuthorizationState {
         ALLOWED,
         NOT_ADMIN,
-        UNKNOWN_USER
+        UNKNOWN_USER,
+        NOT_DELEGATED
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/Authorizator.java:34]: How about: ALLOWED, FORBIDDEN, UNKNOWN_USER ?

File: mailbox/api/src/main/java/org/apache/james/mailbox/exception/UserIsNotDelegatedException.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ ****************************************************************/
+
+package org.apache.james.mailbox.exception;
+
+import org.apache.james.core.Username;
+
+public class UserIsNotDelegatedException extends MailboxException {
+
+    public UserIsNotDelegatedException(Username baseUser, Username nonDelegatedUser) {
+        super("User " + nonDelegatedUser.asString() + " is not delegated by " + baseUser.asString());
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/exception/UserIsNotDelegatedException.java:27]: Why do we need this new exception? 

I would prefer that we reword `NotAdminException` to `ForbiddenDelegationException` to be honnest.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/SessionProviderImpl.java [hunk 3]
```diff
@@ -67,18 +68,20 @@ public MailboxSession login(Username userid, String passwd) throws MailboxExcept
     }
 
     @Override
-    public MailboxSession loginAsOtherUser(Username adminUserid, String passwd, Username otherUserId) throws MailboxException {
-        if (! isValidLogin(adminUserid, passwd)) {
+    public MailboxSession loginAsOtherUser(Username thisUserId, String passwd, Username otherUserId) throws MailboxException {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/SessionProviderImpl.java:71]: :+1:  for this variables renames, it is at least more clear to me now.

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Inject
+    public DelegationStoreAuthorizator(DelegationStore delegationStore, UsersRepository usersRepository) {
+        this.delegationStore = delegationStore;
+        this.usersRepository = usersRepository;
+    }
+
+    @Override
+    public AuthorizationState canLoginAsOtherUser(Username userId, Username otherUserId) throws MailboxException {
+        boolean isAuthorized = Flux.from(delegationStore.authorizedUsers(otherUserId)).collectList().block().contains(userId);
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java:48]: Flux.from(delegationStore.authorizedUsers(otherUserId)).filter(userId::equals).hasElement().block();
 
 Avoid materializing the whole collection.

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizatorTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    void canLoginAsOtherUserShouldReturnAllowedWhenGivenUserIsAdmin() throws Exception {
+        when(usersRepository.isAdministrator(GIVEN_USER))
+            .thenReturn(true);
+        when(usersRepository.contains(OTHER_USER))
+            .thenReturn(true);
+        when(delegationStore.authorizedUsers(OTHER_USER))
+            .thenReturn(Flux.empty());
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizatorTest.java:57]: Hate mocks. Can't we rely on `server-data-memory` instead ? <3

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.james.user.api.DelegationStore;
+import org.apache.james.user.api.UsersRepository;
+import org.apache.james.user.api.UsersRepositoryException;
+
+import reactor.core.publisher.Flux;
+
+/**
+ * Authorizator which use an DelegationStore to check if the delegation is allowed
+ */
+public class DelegationStoreAuthorizator implements Authorizator {
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java:36]: Guice binding for cassandra products maybe?

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractAuthProcessor.java [hunk 7]
```diff
@@ -111,11 +114,14 @@ protected void doAuthWithDelegation(AuthenticationAttempt authenticationAttempt,
...
-            LOGGER.info("User {} does not exist", authenticationAttempt.getAuthenticationId(), e);
+            LOGGER.info("User {} does not exist", otherUser.get(), e);
             no(request, responder, HumanReadableText.USER_DOES_NOT_EXIST);
         } catch (NotAdminException e) {
-            LOGGER.info("User {} is not an admin", authenticationAttempt.getDelegateUserName(), e);
+            LOGGER.info("User {} is not an admin", givenUser, e);
             no(request, responder, HumanReadableText.NOT_AN_ADMIN);
+        } catch (UserIsNotDelegatedException e) {
+            LOGGER.info("User {} is not delegated by {}", givenUser, otherUser.get(), e);
+            no(request, responder, HumanReadableText.NOT_DELEGATED_USER);
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractAuthProcessor.java:124]: These two catch blocks should be one ;-)

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java (hunk 6 above)
Comment (quantranhong1999) [reply server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java:36]: https://github.com/apache/james-project/pull/1032/commits/745bd391d9e55b47eb8699617e7bdf1665ef8cca

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java (hunk 4 above)
Comment (chibenwa) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java:48]: This represent an extra DB read for each and every login attempt. Maybe we need to look up the delegation store if, and only if the two usernames are different?

If they are the same we can skip the delegation and assume ALLOWED here?

Comment (quantranhong1999) [reply server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java:48]: We already checked that using `AuthenticationAttempt::isDelegation` before deciding to do delegation or not. So I think it is no need for further checking.

Comment (chibenwa) [reply server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java]: Ok fair. We will need to keep that in mind for the OIDC delegation stuff, not to call loginAsOtherUser if both users are the same.

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    public DelegationStoreAuthorizator(DelegationStore delegationStore, UsersRepository usersRepository) {
+        this.delegationStore = delegationStore;
+        this.usersRepository = usersRepository;
+    }
+
+    @Override
+    public AuthorizationState canLoginAsOtherUser(Username userId, Username otherUserId) throws MailboxException {
+        boolean isAuthorized = Flux.from(delegationStore.authorizedUsers(otherUserId)).collectList().block().contains(userId);
+        try {
+            if (usersRepository.isAdministrator(userId) || isAuthorized) {
```
Comment (vttranlina) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizator.java:50]:
```suggestion
            if (isAuthorized || usersRepository.isAdministrator(userId)) {
```
When `isAuthorized` is true, Java will not need compute `usersRepository.isAdministrator(userId)`

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractAuthProcessor.java [hunk 9]
```diff
@@ -89,17 +89,19 @@ protected void doAuth(AuthenticationAttempt authenticationAttempt, ImapSession s
...
                 authFailure = true;
             }
             if (!authFailure) {
                 final MailboxManager mailboxManager = getMailboxManager();
                 try {
-                    final MailboxSession mailboxSession = mailboxManager.loginAsOtherUser(authenticationAttempt.getAuthenticationId(),
+                    final MailboxSession mailboxSession = mailboxManager.loginAsOtherUser(givenUser,
                         authenticationAttempt.getPassword(),
-                        authenticationAttempt.getDelegateUserName().get());
+                        otherUser.get());
```
Comment (vttranlina) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractAuthProcessor.java:103]: Why `otherUser` is optional, but here we get it, without empty checking?

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizatorTest.java [hunk 10]
```diff
@@ -43,7 +45,7 @@ class DelegationStoreAuthorizatorTest {
     @BeforeEach
     public void setUp() throws Exception {
         usersRepository = mock(UsersRepository.class);
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/DelegationStoreAuthorizatorTest.java:47]: Idem, do we need to use a mock here?

**Review by chibenwa — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**